File: Solutions/1870. Minimum Speed to Arrive on Time/1870. Minimum Speed to Arrive on Time.py
class Solution:
    def minSpeedOnTime(self, dist: List[int], hour: float) -> int:
        lo,hi,res,n=1,10000000,-1,len(dist)
        while lo<=hi:
            mid=(lo+hi)//2
            s=0.0
            for i in range(n-1):
                s+=math.ceil(dist[i]/mid)
            s+=dist[n-1]/mid
            #print(s)
            if s>hour:
                lo=mid+1
            else:
                hi=mid-1
                res=mid
        return res

File: Solutions/1870. Minimum Speed to Arrive on Time/1870. Minimum Speed to Arrive on Time.java
class Solution {
    public int minSpeedOnTime(int[] dist, double hour) {
        int lo=1,hi=10000000,res=-1,n=dist.length;
        while(lo<=hi){
            int mid=(lo+hi)/2;
            double sum=0;
            for(int i=0;i<n-1;++i)
                sum+=Math.ceil((double)dist[i]/mid);
            sum+=(double)dist[n-1]/mid;
            if(sum>hour)
                lo=mid+1;
            else{
                res=mid;
                hi=mid-1;
            }
        }
        return res;
    }
}